### ISSUE TITLE
./configure fails with Unicode build

### ISSUE THREAD

**Issue by dEajL3kA** [curl/curl#7229]
### I did this

I am trying to build curl with Unicode support on Windows (using MinGW/MSYS2).

In order to enable Unicode I added `-munciode` and `-mconsole` to the CFLAGS, because apparently `./configure` doesn't have a dedicated `--enable-unicode` switch. Or did I miss something?

Unforntunately, this caused `./configure` to fail with "No working C compiler" error. My analysis of the `config.log` showed that `./configure` uses a lot of "test" programs that have `main()` hardcoded, but should be using `wmain()` when Unicode is used! In the actual curl source codes we already *do* have the required `#ifdef`s to use either `main()` or `wmain()` as needed. Only the `./configure` has problem! My workaround was to replace all occurences of "main" in `./configure` with "wmain" using a simple `sed` command. This *does* work. After patching `configure`, Unicode version curl compiled without any problem!

But this workaround is rather cumbersome and other people may run into the same issue. So I think it would be much preferable to have an `--enable-unicode` switch that not only addes `-munciode` and `-mconsole` to the CFLAGS automatically but also uses the "correct" main function in the "test" programs.

### I expected the following

`CFLAGS="-municode -mconsole" ./configure && make` should just work *without* the need to patch `configure`.

Even better, `./configure --enable-unicode && make` should be available.

### curl/libcurl version

7.77.0

### operating system

Windows 10, using latest MinGW/MSYS2 compiler + build environment

**Comment by jay** [comment]
Can you use -DUNICODE -D_UNICODE instead

**Comment by dEajL3kA** [comment]
> Can you use -DUNICODE -D_UNICODE instead

Sorry, but no. Or I am missing something 😕 

* To the best of my understanding, `-municode -mconsole` already implies `-DUNICODE -D_UNICODE`, so explicitely adding `-DUNICODE -D_UNICODE` when `-municode -mconsole` is already set does **not** change anything.

* Setting ***only*** `-DUNICODE -D_UNICODE` to CFLAGS but **not** `-municode -mconsole` will cause MinGW/GCC to link against `main()` rather than `wmain()`. As a result, `./configure` now *does* successe without additional hacks, but later the linking of `curl.exe` **fails** with undefinied reference error – which is kind of expected, because with  `-D_UNICODE` we tell `tool_main.c` to generate `wmain()` instead of `main` while at the same time MinGW/GCC is expecting a `main()` function due to the lack of `-municode -mconsole`.

**Comment by jay** [comment]
It seems [this](https://github.com/curl/curl-for-win/blob/559ca8b2adf594a33c3128e682be3b53f48e0b73/curl.sh#L53-L54) is the way curl-for-win does it:

~~~shell
  CURL_CFLAG_EXTRAS="${CURL_CFLAG_EXTRAS} -DUNICODE -D_UNICODE"
  CURL_LDFLAG_EXTRAS_EXE="${CURL_LDFLAG_EXTRAS_EXE} -municode"
~~~

So we'd have to apply -municode only for the curl tool?

/cc @vszakats

**Comment by dEajL3kA** [comment]
> So we'd have to apply -municode only for the curl tool?

Yes, I think so. But it seems that `CURL_LDFLAG_EXTRAS_EXE` does *nothing* for me. I did a full-text search on both, `configure` and the resulting `Makefile`, but could **not** find the identifier `CURL_LDFLAG_EXTRAS_EXE` at all.

However, what ***does*** work for me is:
```
CFLAGS="-DUNICODE -D_UNICODE" ./configure
make curl_LDFLAGS="-all-static -municode -mconsole"
```

**Comment by jay** [comment]
> Maybe add a `./configure` option for Unicode build?

Seems like a good idea.

> (BTW: What is the difference between `CURL_CFLAG_EXTRAS` and `CFLAGS` variables?)

I think it was added to pass cflags without overriding the default CFLAGS. It's not documented and I may be wrong...

**Comment by vszakats** [comment]
`CURL_LDFLAG_EXTRAS_EXE` is `Makefile.m32` specific, so it indeed doesn't do anything with ./Configure builds.

**Comment by bagder** [comment]
This issue seems to have stalled. Is there anything we can/should do about this report? Any proposed next steps?

**Comment by vszakats** [comment]
I think it'd be best to complete Unicode support in libcurl and then fix the different build systems to enable it correctly. Though these two can be done in any order or in parallel. @jay had a nice patch for the former, though I continue to think that fall-back methods are bad and should not be used in libcurl. (Yes, OpenSSL uses one such method, but OpenSSL also keeps at least one Windows vulnerability open since years without intention to fix it (along some other past woes on Windows), so I think it should not be regarded as the yardstick for Windows-development specifically.)

**Comment by dEajL3kA** [comment]
> I think it'd be best to complete Unicode support in libcurl and then fix the different build systems to enable it correctly. Though these two can be done in any order or in parallel. @jay had a nice patch for the former, though I continue to think that fall-back methods are bad and should not be used in libcurl. 

👍 ⚡

**Comment by Thanayaby** [comment]
Nothing

**Comment by bagder** [comment]
Anyone care to write this issue up in a brief summary for adding to the `KNOWN_BUGS` document?

**Comment by jay** [comment]
IMO this isn't a bug, it's either an enhancement or a feature request. The user wants an easier way to tell autotools to build curl with Windows Unicode support, like `./configure --enable-windows-unicode`

**Comment by vszakats** [comment]
I tried to fix this without the `sed` trick, and tested: `-Dmain=wmain`.
Oddly enough, it worked!

```shell
CPPFLAGS="${CPPFLAGS} -DUNICODE -D_UNICODE"
CPPFLAGS="${CPPFLAGS} -Dmain=wmain"
LDFLAGS="${LDFLAGS} -municode"
```

It'd be better to make `-Dmain=wmain` internal to `./configure`, if possible.

If this is a feasible solution, adding a `--enable-windows-unicode` is just the icing on the cake.

There is this warning, unrelated to either of the above hacks:
```
../../src/tool_main.c:239:5: warning: no previous prototype for function 'wmain' [-Wmissing-prototypes]
int wmain(int argc, wchar_t *argv[])
    ^
1 warning generated.
```

**Comment by lordmulder** [comment]
Just tried the new configure option in 8.10.0.

Still same old problem: Configure script doesn't find "main" function, because, in Windows, when Unicode is enabled, the main function **must** be `wmain()` rather `main()`, but it's apparently hard-coded to `main()` in your configure script:

```
C:/M/B/src/mingw-w64/mingw-w64-crt/crt/crtexewin.c:67:(.text+0xc1): undefined reference to `wWinMain@16'
collect2.exe: error: ld returned 1 exit status
configure:5018: $? = 1
configure:5058: result: no
configure: failed program was:
| /* confdefs.h */
| #define PACKAGE_NAME "curl"
| #define PACKAGE_TARNAME "curl"
| #define PACKAGE_VERSION "-"
| #define PACKAGE_STRING "curl -"
| #define PACKAGE_BUGREPORT "a suitable curl mailing list: https://curl.se/mail/"
| #define PACKAGE_URL ""
| /* end confdefs.h.  */
| 
| int main (void)
| {
| 
|  ;
|  return 0;
| }
configure:5063: error: in `/d/_Sandbox/curl/build/x86/curl':
configure:5065: error: C compiler cannot create executables
See `config.log' for more details
```

**Comment by vszakats** [comment]
We have a CI job that builds and tests fine without this hack:
https://github.com/curl/curl/actions/runs/10835445722/job/30067013990#step:8:936

I've found that `wmain` is only necessary when linking with `-municode`. The merged implementation does that only for the curl tool, so everything else continues to build with `main`, as before.

The hack also doesn't address building curl tests, which also all use `main`.

**Comment by lordmulder** [comment]
> I've found that wmain is only necessary when linking with `-municode`

Yeah, but `-municode` must be added to `LDFLAGS` in order to produce a working Unicode binary.

If `-municode` is **not** added to `LDFLAGS`, then the normal `main()` function will be used, which might avoid the linking issue with the Configure script, but also means that the command-line will be mangled (i.e. any non-Unicode character is replaced by a `?` character) even before it arrives in "main" function. So, even if all the rest was compiled as Unicode, we have a problem!

To actually get the untainted command-line into "main" as Unicode, we need to use the `wmain()` function, which requires  `-municode` to be added to `LDFLAGS`, which then breaks the Configure script (as it still expects `main()` function).

**Comment by vszakats** [comment]
> > I've found that wmain is only necessary when linking with `-municode`
> 
> Yeah, that's right. But `-municode` **must** be added to `LDFLAGS` in order to produce a working Unicode binary.

As mentioned in the previous post, `-municode` is added for the curl tool automatically. The curl tool does use `wmain()`:
https://github.com/curl/curl/blob/8ca60308398e42d77d140ebe5c3331c6e086d355/src/tool_main.c#L218-L228
https://github.com/curl/curl/blob/8ca60308398e42d77d140ebe5c3331c6e086d355/src/Makefile.am#L54-L56

[ Generally speaking: `-municode` controls the app startup code, and Unicode app code can live without it if it doesn't want to get a wide argv via main. There are other means to retrieve the command-line in Unicode too. So no, `-municode` isn't _required_ for a Unicode app. The curl tool uses it, the feature tests and regression tests don't. ]

> (I posted the result when `-municode` is used, but Configure it is **not** manually patched [above](https://github.com/curl/curl/issues/7229#issuecomment-2347308213). As you can see, it's failing right away. Currently, passing the new option `--enable-windows-unicode` apparently does _nothing_ to fix this issue)

You seems to be passing `LDFLAGS=-municode` to the build and that is what breaks it for you. The solution is to drop this option and use `--enable-windows-unicode` only.

Have you tried such build?

**Comment by lordmulder** [comment]
Ah, I see now. You mean we can leave out `-municode` from the `LDFLAGS` for the `./configure` invocation now (something the was absolutely required in the past!) and then `--enable-windows-unicode` will sneak it in again, but _only_ for linking the final cURL executable? Will have to give this a try, and see if there are any undesired side effects with any of the dependencies...

**Comment by lordmulder** [comment]
> Ah, I see now. You mean we can leave out `-municode` from the `LDFLAGS` for the `./configure` invocation now (something that was absolutely required in the past!) and then `--enable-windows-unicode` will sneak it in again, but _only_ for linking the final cURL executable? Will have to give this a try, and see if there are any undesired side effects with any of the dependencies...

Indeed, seems to work 🥳
